src/shell/shell_history.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ *  Copyright (C) 2023-2023  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "shell.h"
+
+#include <fstream>
+
+#include "checks.h"
+#include "control.h"
+#include "dosbox.h"
+#include "fs_utils.h"
+#include "string_utils.h"
+
+CHECK_NARROWING();
+
+static constexpr int HistoryMaxLineSize = 256;
+static constexpr int HistoryMaxNumLines = 500;
+
+static std_fs::path get_shell_history_path();
+static bool command_is_exit(std::string_view command);
+
+ShellHistory& ShellHistory::Instance()
+{
+	static auto history = ShellHistory();
+	return history;
+}
+
+void ShellHistory::Append(std::string command, std::uint16_t code_page)
+{
+	auto to_utf8_str = [code_page](const std::string& dos_str) {
+		std::string utf8_str{};
```

Comment:
As discussed previously, please use the `= {}` initialiser format everywhere as the project conventions.